### PR TITLE
Project inception

### PR DESCRIPTION
**Pull request by ElChache**
First release with the following functionality:

- Two types of steps: resolvers, that augment the context, and taps, that dont', but can still short-circuit
- Global taps: exception tap and anomaly tap
- 3 callbacks: on-complete, on-anomaly, on-error
- An optional anomaly? on the configuration that allows to use custom anomalies
- log-step-fn, that allows to customize how each step gets logged

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,96 @@
...
+      - [opt] log-step-fn   A function that defines how each step adds information to the log
+
+- **steps**: Each item on the steps collection must be either a TapStep, or a ResolverStep
+
+      TapStep:
+       - tap:  A function that gets the context but doesn't augment it
+       - name: The name of the step
+
+      ResolverStep:
+       - resolver: A function that gets the context and assocs the result into it on the given path
```
Comment (arichiardi) [inline README.md]: Not convinced about this name, @mattbishop do you have a better name?

File: README.md [hunk 2]
```diff
@@ -1,2 +1,96 @@
...
+ 
+
+
+## Example
+
+```clojure
+(fonda/execute
+  {:exception-tap (fn [{:keys [error]}] (js/console.log "Exception happened:" error))
+   :anomaly-tap   (fn [{:keys [anomaly]}] (js/console.log "An anomaly happened:" anomaly))}
+  [{:path [:remote-thing] :resolver (fn [ctx] (ajax/get "http://remote-thing-url"))}
```
Comment (arichiardi) [inline README.md]: just for clarity this can be on a separate line with spaces?

File: README.md [hunk 3]
```diff
@@ -1,2 +1,96 @@
...
+   {:tap (fn [{:keys [remote-thing]}] (js/console.log "remote thing:" ()))}
+   {:path [:remote-thing-processed] :resolver (fn [{:keys [remote-thing]} (process-remote-thing remote-thing)])}]
+
+  (fn [{:keys [remote-thing-processed]}] remote-thing-processed)
+  (fn [anomaly] anomaly)
```
Comment (arichiardi) [inline README.md]: I would add some side effect here just for clarity, say calling a function?

Comment (arichiardi) [reply README.md]: Like `(fn [anomaly] (do-something-with-anomaly ...))`

File: deps.edn [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{:paths   ["src"]
+ :deps    {com.bhauman/spell-spec {:mvn/version "0.1.1"}}
+ :aliases {:test {:extra-paths ["test"]
+                  :extra-deps  {}}
+           :dev  {:extra-paths ["dev"]
+                  :extra-deps  {org.clojure/clojure       {:mvn/version "1.10.0-beta5"}
+                                org.clojure/clojurescript {:mvn/version "1.10.339"}
```
Comment (arichiardi) [inline deps.edn]: Let's bump ClojureScript and shadow

File: deps.edn [hunk 5]
```diff
@@ -0,0 +1,10 @@
+{:paths   ["src"]
+ :deps    {com.bhauman/spell-spec {:mvn/version "0.1.1"}}
```
Comment (arichiardi) [inline deps.edn]: Do we use this? If not we can remove it?

File: project.clj [hunk 6]
```diff
@@ -0,0 +1,11 @@
+(defproject fonda "0.0.1"
```
Comment (arichiardi) [inline project.clj]: let's add here the package group like `elasticpath/fonda`

File: src/fonda/anomaly.cljs [hunk 7]
```diff
@@ -0,0 +1,6 @@
+(ns fonda.anomaly)
+
+(def anomaly-key :cognitect.anomalies/anomaly)
+
+(defn anomaly? [m]
+  (some? (get m anomaly-key)))
```
Comment (arichiardi) [inline src/fonda/anomaly.cljs]: I would put the key inline, not a big deal though

File: src/fonda/core.cljs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  - `on-exception` Callback that gets called with an exception when any step triggers one
+  "
+  ([config steps ctx-seed on-complete on-anomaly on-exception]
+   (let [{:keys [anomaly?
+                 exception-tap
+                 anomaly-tap
+                 log-step-fn]} config]
+
+     (-> (r/map->RuntimeContext
+           {:anomaly?      (or anomaly? fonda.anomaly/anomaly?)
```
Comment (arichiardi) [inline src/fonda/core.cljs]: Nice!

File: README.md [hunk 9]
```diff
@@ -1,2 +1,96 @@
...
+      - [opt] anomaly-tap   A function that gets called with te runtime-context when a step returns an anomaly
+      - [opt] log-step-fn   A function that defines how each step adds information to the log
+
+- **steps**: Each item on the steps collection must be either a TapStep, or a ResolverStep
+
+      TapStep:
+       - tap:  A function that gets the context but doesn't augment it
+       - name: The name of the step
+
+      ResolverStep:
```
Comment (mattbishop) [inline README.md]: Might call it just `step`?

File: README.md [hunk 10]
```diff
@@ -1,2 +1,96 @@
...
+(execute config steps ctx-seed on-complete on-anomaly on-exception)
+```
+- **config** A map with:
+
+      - [opt] anomaly?      A function that gets a map and determines if it is an anomaly
+      - [opt] exception-tap A function gets called with the runtime-context when there is an exception
+      - [opt] anomaly-tap   A function that gets called with te runtime-context when a step returns an anomaly
+      - [opt] log-step-fn   A function that defines how each step adds information to the log
+
+- **steps**: Each item on the steps collection must be either a TapStep, or a ResolverStep
```
Comment (mattbishop) [inline README.md]: ...either a Tap or a Step`

File: README.md [hunk 11]
```diff
@@ -1,2 +1,96 @@
...
+      TapStep:
+       - tap:  A function that gets the context but doesn't augment it
+       - name: The name of the step
+
+      ResolverStep:
+       - resolver: A function that gets the context and assocs the result into it on the given path
+       - path:     Path where to assoc the result of the resolver
+       - name:     The name of the step
+       
+- **ctx-seed** The context data that gets passed to the steps functions.
```
Comment (mattbishop) [inline README.md]: `seed` makes me think random numbers and cryptography. How about **initial-ctx**

File: src/fonda/runtime.cljs [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+   on-anomaly
+
+   ;; The steps that haven't been already processed
+   queue
+
+   ;; A log, each step can add information here
+   step-log
+
+   ;; A function that defines how each step adds information to the log
+   log-step-fn])
```
Comment (arichiardi) [inline src/fonda/runtime.cljs]: I don't know if we want this actually

Comment (arichiardi) [reply src/fonda/runtime.cljs]: I mean not customizable from the outside

File: README.md [hunk 13]
```diff
@@ -1,2 +1,96 @@
...
+               Must be either a map, or nil
```
Comment (mattbishop) [inline README.md]: Perhaps the default is map. Never nil.

File: test/fonda/core_test.cljs [hunk 14]
```diff
@@ -0,0 +1,465 @@
...
+(orchestra/instrument)
+
+(defn anomaly
+  ([category]
+   {fonda.anomaly/anomaly-key {:category category}})
+  ([category message]
+   {fonda.anomaly/anomaly-key {:category category
+                               :message  message}}))
+
+#_(execute
```
Comment (arichiardi) [inline test/fonda/core_test.cljs]: leftover

File: README.md [hunk 15]
```diff
@@ -1,2 +1,96 @@
...
+               
+- **on-complete**  Callback that gets called with the context if all the steps succeeded
+- **on-anomaly**   Callback that gets called with an anomaly when any step returns one
+- **on-exception** Callback that gets called with an exception when any step triggers one
+
+### Steps
+
+There are two types of steps: **taps** and **resolvers**.
```
Comment (mattbishop) [inline README.md]: Maybe `taps` and `processors`

File: README.md (hunk 11 above)
Comment (arichiardi) [reply README.md]: yeah that was taken from the `reduce` literature...usually the use a `seed`, but `initial` sounds good as well

File: README.md (hunk 15 above)
Comment (arichiardi) [reply README.md]: I like processors

File: README.md (hunk 2 above)
Comment (mattbishop) [reply README.md]: Yes good call. With multiline statements too.

Comment (mattbishop) [reply README.md]: Also why is line 88-90 a vector?

File: README.md (hunk 9 above)
Comment (arichiardi) [reply README.md]: So steps have processors, sounds good

File: README.md (hunk 1 above)
Comment (ElChache) [reply README.md]: processor will be!

File: README.md (hunk 15 above)
Comment (ElChache) [reply README.md]: processors it is!

File: README.md [hunk 16]
```diff
@@ -1,2 +1,96 @@
...
+Fonda sequentially executes an (a)synchronous series of steps, one after the other, augmenting a context map.
+
+## Syntax
+
+```clojure
+(execute config steps ctx-seed on-complete on-anomaly on-exception)
+```
+- **config** A map with:
+
+      - [opt] anomaly?      A function that gets a map and determines if it is an anomaly
```
[the remark this answers is not in the document]
Comment (arichiardi) [reply README.md]: Agree

File: README.md [hunk 17]
```diff
@@ -1,2 +1,96 @@
...
+      - [opt] exception-tap A function gets called with the runtime-context when there is an exception
+      - [opt] anomaly-tap   A function that gets called with te runtime-context when a step returns an anomaly
```
[the remark this answers is not in the document]
Comment (ElChache) [reply README.md]: They are called before the callbacks, so it is not really the last thing that happens

File: README.md [hunk 18]
```diff
@@ -1,2 +1,96 @@
...
+      - [opt] log-step-fn   A function that defines how each step adds information to the log
```
[the remark this answers is not in the document]
Comment (ElChache) [reply README.md]: It could be useful for debugging to see what steps have run. For now it just adds step names to a vector. It would be possible to pass your own thing here. But if you guys think that is not needed we can remove it from the API.

File: src/fonda/runtime.cljs (hunk 12 above)
Comment (ElChache) [reply src/fonda/runtime.cljs]: I'll remove it from the API if you guys want this out. To me, personally, it doesn't bother me as it is optional, and I could imagine a situation where I benefit from more sophisticated login (showing the diffs on the context on every step could be a nice one)

File: test/fonda/core_test.cljs (hunk 14 above)
Comment (ElChache) [reply test/fonda/core_test.cljs]: Moved to the README

File: README.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (ElChache) [reply README.md]: Added an explanation on the readme, and a link to the cognitect anomalies

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ElChache) [reply README.md]: You are so right! Fixed

File: README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (ElChache) [reply README.md]: yes, initial-ctx now

File: README.md [hunk 19]
```diff
@@ -1,2 +1,96 @@
...
+
+      ResolverStep:
+       - resolver: A function that gets the context and assocs the result into it on the given path
+       - path:     Path where to assoc the result of the resolver
+       - name:     The name of the step
+       
+- **ctx-seed** The context data that gets passed to the steps functions.
+               Must be either a map, or nil
+               
+- **on-complete**  Callback that gets called with the context if all the steps succeeded
```
[the remark this answers is not in the document]
Comment (ElChache) [reply README.md]: Not so sure what -the you mean here

File: README.md [hunk 20]
```diff
@@ -1,2 +1,96 @@
...
+- **on-anomaly**   Callback that gets called with an anomaly when any step returns one
+- **on-exception** Callback that gets called with an exception when any step triggers one
+
+### Steps
+
+There are two types of steps: **taps** and **resolvers**.
+
+Resolvers get a path and a function. The data returned will be placed on the context on that path.
```
[the remark this answers is not in the document]
Comment (ElChache) [reply README.md]: yep! fixed

File: README.md [hunk 21]
```diff
@@ -1,2 +1,96 @@
...
+
+Taps only get a function, and if they succeed, the result is ignored.
+
+If any step returns an anomaly, or triggers an exception, the execution of the steps stops and the global taps and
```
[the remark this answers is not in the document]
Comment (ElChache) [reply README.md]: thanks man, I appreciate that you actually read it! Fixed!

File: README.md (hunk 13 above)
Comment (ElChache) [reply README.md]: not nil anymore

File: README.md [hunk 22]
```diff
@@ -1,2 +1,96 @@
...
+If any step returns an anomaly, the anomaly-tap will be called with the RuntimeContext, and then the on-anomaly callback
+
+If any step triggers an exception, the exception-tap will be called with the RuntimeContext, and then on-exception callback.
+
+#### Resolver steps
+
+A resolvers are maps with the following keys:
+
+- **resolver** A function that gets a context map and returns data. Can be asynchronous.
+               If it returns an anomaly, or it triggers an error, the steps execution will be interrupted.
```
[the remark this answers is not in the document]
Comment (ElChache) [reply README.md]: yes, better expressed that way. Fixed!

File: src/fonda/runtime.cljs (hunk 12 above)
Comment (ElChache) [reply src/fonda/runtime.cljs]: Ok, removed it from the api.

File: README.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (ElChache) [reply README.md]: Yes so I see you guys didn't like this much. I removed it from the api. Internally still logs the steps names in a log

File: README.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (yilazius) [reply README.md]: I meant change to `if all steps succeeded`

File: CHANGELOG.md [hunk 23]
```diff
@@ -0,0 +1,12 @@
...
+
+All notable changes to this project will be documented in this file.
+
+The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
+and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
+
+## [master](https://github.com/elasticpath/terraform-vars/compare/ece2cb8...HEAD) (unreleased)
+
+### Added
+
```
Comment (arichiardi) [inline CHANGELOG.md]: Can we add the project inception commit here?

File: README.md [hunk 24]
```diff
@@ -1,2 +1,160 @@
 # fonda
```
Comment (arichiardi) [inline README.md]: Can we add our wonderful logo on the top?

```
<img src="https://www.elasticpath.com/sites/all/themes/bootstrap/images/elastlic-path-logo-RGB.svg" alt="elasticpath logo" title="elasticpath" align="right" width="150"/>
```

File: README.md [hunk 25]
```diff
@@ -1,2 +1,160 @@
...
+
 An async pipeline approach to functional core - imperative shell.
+
+## Asynchronous pipeline of steps
+
+Fonda sequentially executes an (a)synchronous series of steps, one after the other, augmenting a context map.
+
+Fonda distinguishes between exception errors and [anomaly](https://github.com/cognitect-labs/anomalies/blob/master/src/cognitect/anomalies.cljc) errors
```
Comment (arichiardi) [inline README.md]: I would end up with `and anomalies`, cause error is already clear and muddies the vocabulary a bit

File: README.md [hunk 26]
```diff
@@ -1,2 +1,160 @@
...
+
+Anomalies are just data that define an abnormally. By default, Fonda wraps the anomalies in a map with the key :anomaly.
```
Comment (arichiardi) [inline README.md]: I would rephrase it as:

> Anomalies are just data that define a recoverable "business" error, as opposed to JavaScript `js/Error` or Java `Exception`, which never meant to be caught and are usually caused by programming bugs.

Also I think you want to write the fully qualified name there, which is `:cognitect.anomalies/anomaly`

File: README.md [hunk 27]
```diff
@@ -1,2 +1,160 @@
...
+
+## Syntax
+
+```clojure
+(execute config steps initial-ctx on-complete on-anomaly on-exception)
+```
+- **config** A map with:
+
+      - [opt] anomaly?      A function that gets a map and determines if it is an anomaly
+      - [opt] exception-tap A function gets called with the runtime-context when there is an exception.
```
Comment (arichiardi) [inline README.md]: I think runtime context here is two words

File: README.md [hunk 28]
```diff
@@ -1,2 +1,160 @@
...
+
+### Steps
+
+There are two types of steps: **taps** and **processors**.
+
+Processors get a path and a function. The data returned will be placed in the context on that path.
+
+Taps only get a function, and if they succeed, the result is ignored.
+
+If any step returns an anomaly, or triggers an exception, the execution of the steps stops and the global taps
```
Comment (arichiardi) [inline README.md]: This is the core of our semantics so I would put it in a nice formatted bulletted list :smile:

File: README.md [hunk 29]
```diff
@@ -1,2 +1,160 @@
...
+      Tap:
+       - tap:  A function that gets the context but doesn't augment it
+       - name: The name of the step
+
+      Processor:
+       - processor: A function that gets the context returns a result that is assoced into the context on the given path
+       - path:      Path where to assoc the result of the processor
+       - name:      The name of the step
+       
+- **initial-ctx** The context data that gets passed to the steps functions. Must be a map
```
Comment (arichiardi) [inline README.md]: Maybe we can use `:initial-ctx` (which a colon) cause those are keys

File: shadow-cljs.edn [hunk 30]
```diff
@@ -0,0 +1,20 @@
+{:deps   {:aliases [:dev :test]}
+ :builds {
+          :lib  {:target    :node-library
```
Comment (arichiardi) [inline shadow-cljs.edn]: I think we can remove this section for now as we are not going to deploy it as JavaScript

File: src/fonda/core.cljs [hunk 31]
```diff
@@ -0,0 +1,75 @@
...
+               :initial-ctx ::r/ctx
+               :on-complete ::r/on-complete
+               :on-anomaly ::r/on-anomaly
+               :on-exception ::r/on-exception))
+(defn execute
+  "Sequentially executes the series of given `steps`
+  Each step tap of processor function can be synchronous or asynchronous.
+  .
+
+  - `config`: A map with:
```
Comment (arichiardi) [inline src/fonda/core.cljs]: We can use colons here too

File: src/fonda/execute.cljs [hunk 32]
```diff
@@ -0,0 +1,144 @@
...
+  :args (s/cat :runtime-ctx ::r/runtime-context))
+
+(defn set-tap-result [{:as runtime-ctx :keys [anomaly?]} res]
+  (if (anomaly? res)
+    (assoc runtime-ctx :anomaly res)
+    runtime-ctx))
+
+
+
+(s/fdef set-processor-result
```
Comment (arichiardi) [inline src/fonda/execute.cljs]: why no `assoc`? Set is so OO :smile:

File: src/fonda/runtime.cljs [hunk 33]
```diff
@@ -0,0 +1,70 @@
...
```
Comment (arichiardi) [inline src/fonda/runtime.cljs]: This is gone isn't it?

File: README.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (ElChache) [reply README.md]: Added two lines explaining how the loggers are called after the steps, and then the callbacks

File: README.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (ElChache) [reply README.md]: yes, I don't know why I felt the need for "the" there

File: CHANGELOG.md (hunk 23 above)
Comment (ElChache) [reply CHANGELOG.md]: sure, and also a summary of the features

File: README.md (hunk 25 above)
Comment (ElChache) [reply README.md]: sure, I took the "error" word out of there

File: README.md (hunk 26 above)
Comment (ElChache) [reply README.md]: sure, replaced with yours and with the fully qualified key

File: README.md (hunk 28 above)
Comment (ElChache) [reply README.md]: Bulleted!

File: shadow-cljs.edn (hunk 30 above)
Comment (ElChache) [reply shadow-cljs.edn]: cuz of the REPL! :p

File: src/fonda/core.cljs (hunk 31 above)
Comment (ElChache) [reply src/fonda/core.cljs]: I added colons in all the keywords, but here they are positional arguments

File: src/fonda/execute.cljs (hunk 32 above)
Comment (ElChache) [reply src/fonda/execute.cljs]: you are so right my friend

File: src/fonda/runtime.cljs (hunk 33 above)
Comment (ElChache) [reply src/fonda/runtime.cljs]: It's gone from the api but still used internally to add the step names into the step_log

File: README.md (hunk 27 above)
Comment (ElChache) [reply README.md]: right right

File: README.md (hunk 29 above)
Comment (ElChache) [reply README.md]: Not a key on the api, it is a positional argument

File: README.md [hunk 34]
```diff
@@ -1,2 +1,176 @@
...
+   ;; Processor that processes the data
+   {:processor        (fn [{:keys [remote-thing]}] (process-remote-thing remote-thing))
+    :name             "process-thing"
+    :path             [:remote-thing-processed]}
+   
+   ;; Tap that logs things
+   {:tap              (fn [{:keys [remote-thing]}] (js/console.log "remote thing:"))}
+   ]
+
+  ;; On complete
```
Comment (arichiardi) [inline README.md]: Now it is `On success`

File: README.md [hunk 35]
```diff
@@ -1,2 +1,176 @@
...
+Global taps are called with the runtime context. The runtime context is a map that contains, among other things, 
+the context that is passed to each step. It also contains:
+
+- **:step-log** A collection of step logs. By default only step names are logged.
+
+
+## Examples
+
+```clojure
+(fonda/execute
```
Comment (arichiardi) [inline README.md]: I would keep only the filled in one

File: README.md [hunk 36]
```diff
@@ -1,2 +1,176 @@
...
+
+    ;; exception cb
+    (fn [exception])
+
+    )
+```
+
+```clojure
+(fonda/execute
+  {:log-exception     (fn [{:keys [ctx error step-log]}] (js/console.log "Exception happened:" error))
```
Comment (arichiardi) [inline README.md]: maybe the `js.console.log` line can go down one line?

File: README.md (hunk 34 above)
Comment (ElChache) [reply README.md]: good catch!

File: README.md (hunk 35 above)
Comment (ElChache) [reply README.md]: yes, removed the other one
